File: traefik.py
import grafanalib.core as G
from grafanalib import formatunits as UNITS

from helpers import GolangResourceUsagePanels, QueryExpr, RowPanel, StatPanel, TimeSeriesPanel, auto_layout, avg_duration_query, generate_dashboard


def cleanup_service_label(query):
    return f'label_replace({query}, "service", "$1", "service", "([^-]+-[^@]+)@.*")'


error_thresholds = [G.Threshold("green", 0, 0.0), G.Threshold("red", 1, 1.0)]
my_dashboard = G.Dashboard(
    title="Traefik",
    uid="traefik",
    description="Dashboard for Traefik",
    tags=["traefik"],
    timezone="browser",
    refresh=None,
    templating=G.Templating(
        list=[
            G.Template(type="datasource", name="datasource", label="Datasource", query="prometheus"),
            G.Template(type="query", dataSource="${datasource}", name="namespace", label="Namespace", query="label_values(traefik_config_last_reload_success, namespace)", includeAll=True, refresh=G.REFRESH_ON_TIME_RANGE_CHANGE, multi=False, sort=G.SORT_ALPHA_DESC),
            G.Template(type="query", dataSource="${datasource}", name="instance", label="Instance", query='label_values(traefik_config_last_reload_success{namespace=~"$namespace"}, instance)', includeAll=True, refresh=G.REFRESH_ON_TIME_RANGE_CHANGE, multi=True, sort=G.SORT_ALPHA_DESC),
            G.Template(type="query", dataSource="${datasource}", name="entrypoint", label="Entrypoint", query='label_values(traefik_entrypoint_open_connections{namespace=~"$namespace"}, entrypoint)', includeAll=True, refresh=G.REFRESH_ON_TIME_RANGE_CHANGE, multi=True, sort=G.SORT_ALPHA_DESC),
            G.Template(type="query", dataSource="${datasource}", name="service", label="Service", query='label_values(traefik_service_open_connections{namespace=~"$namespace"}, service)', includeAll=True, refresh=G.REFRESH_ON_TIME_RANGE_CHANGE, multi=True, sort=G.SORT_ALPHA_DESC),
        ]
    ),
    panels=auto_layout(
        [
            *RowPanel(
                title="Overview",
                defaultWidth=4,
                panels=[
                    [
                        # StatPanel("Request Count", [QueryExpr('sum(increase(traefik_entrypoint_requests_total{namespace=~"$namespace", instance=~"$instance", entrypoint=~"$entrypoint"}[$__range]))')], unit=UNITS.SHORT),
                        # StatPanel("Request Bytes", [QueryExpr('sum(increase(traefik_entrypoint_requests_bytes_total{namespace=~"$namespace", instance=~"$instance", entrypoint=~"$entrypoint"}[$__range]))')], unit=UNITS.BYTES),
                        # StatPanel("Response Bytes", [QueryExpr('sum(increase(traefik_entrypoint_responses_bytes_total{namespace=~"$namespace", instance=~"$instance", entrypoint=~"$entrypoint"}[$__range]))')], unit=UNITS.BYTES),
                        StatPanel("Request Count", [QueryExpr('sum(increase(traefik_service_requests_total{namespace=~"$namespace", instance=~"$instance", service=~"$service"}[$__range]))')], unit=UNITS.SHORT),
                        StatPanel("Request Bytes", [QueryExpr('sum(increase(traefik_service_requests_bytes_total{namespace=~"$namespace", instance=~"$instance", service=~"$service"}[$__range]))')], unit=UNITS.BYTES),
                        StatPanel("Response Bytes", [QueryExpr('sum(increase(traefik_service_responses_bytes_total{namespace=~"$namespace", instance=~"$instance", service=~"$service"}[$__range]))')], unit=UNITS.BYTES),
                        StatPanel("Config reloads", [QueryExpr('sum(increase(traefik_config_reloads_total{namespace=~"$namespace", instance=~"$instance"}[$__range]))')], unit=UNITS.SHORT),
                        StatPanel("Config reloads failures", [QueryExpr('sum(increase(traefik_config_reloads_failure_total{namespace=~"$namespace", instance=~"$instance"}[$__range]))')], unit=UNITS.SHORT),
                        StatPanel("Last successful config reload", [QueryExpr('max(traefik_config_last_reload_success{namespace=~"$namespace", instance=~"$instance"}[$__range]) * 1000')], unit=UNITS.DATE_TIME_FROM_NOW),
                        StatPanel("Last failed config reload", [QueryExpr('max(traefik_config_last_reload_failure{namespace=~"$namespace", instance=~"$instance"}[$__range]) * 1000')], unit=UNITS.DATE_TIME_FROM_NOW),
                    ],
                ],
            ),
            *RowPanel(
                title="Entrypoint Metrics",  # available labels: entrypoint, code, method, protocol
                defaultWidth=6,
                panels=[
                    [
                        TimeSeriesPanel("Open connections by protocol", [QueryExpr('sum(traefik_entrypoint_open_connections{namespace=~"$namespace", instance=~"$instance", entrypoint=~"$entrypoint"}[$__interval]) by (entrypoint, protocol)', "{{ entrypoint }} - {{ protocol }}")], unit=UNITS.SHORT, w=12),
                        TimeSeriesPanel("Open connections by method", [QueryExpr('sum(traefik_entrypoint_open_connections{namespace=~"$namespace", instance=~"$instance", entrypoint=~"$entrypoint"}[$__interval]) by (entrypoint, method)', "{{ entrypoint }} - {{ method }}")], unit=UNITS.SHORT, w=12),
                        # by entrypoint
                        TimeSeriesPanel("Request Count", [QueryExpr('sum(increase(traefik_entrypoint_requests_total{namespace=~"$namespace", instance=~"$instance", entrypoint=~"$entrypoint"}[$__interval])) by (entrypoint)', "{{ entrypoint }}")], unit=UNITS.SHORT, style="bars"),
                        TimeSeriesPanel("Request Bytes", [QueryExpr('sum(increase(traefik_entrypoint_requests_bytes_total{namespace=~"$namespace", instance=~"$instance", entrypoint=~"$entrypoint"}[$__interval])) by (entrypoint)', "{{ entrypoint }}")], unit=UNITS.BYTES, style="bars"),
                        TimeSeriesPanel("Response Bytes", [QueryExpr('sum(increase(traefik_entrypoint_responses_bytes_total{namespace=~"$namespace", instance=~"$instance", entrypoint=~"$entrypoint"}[$__interval])) by (entrypoint)', "{{ entrypoint }}")], unit=UNITS.BYTES, style="bars"),
                        TimeSeriesPanel("Avg request duration", [QueryExpr(cleanup_service_label(avg_duration_query("traefik_entrypoint_request_duration_seconds", '{namespace=~"$namespace", instance=~"$instance", entrypoint=~"$entrypoint"}', "entrypoint")), "{{ entrypoint }}")], unit=UNITS.SECONDS),
                    ],
                ],
            ),
            *RowPanel(
                title="Service Metrics",  # available labels: service, code, method, protocol
                defaultWidth=6,
                panels=[
                    [
                        TimeSeriesPanel("Open connections by service", [QueryExpr(cleanup_service_label('sum(traefik_service_open_connections{namespace=~"$namespace", instance=~"$instance", service=~"$service"}[$__interval]) by (service)'), "{{ service }}")], unit=UNITS.SHORT, w=8),
                        TimeSeriesPanel("Open connections by protocol", [QueryExpr(cleanup_service_label('sum(traefik_service_open_connections{namespace=~"$namespace", instance=~"$instance", service=~"$service"}[$__interval]) by (protocol)'), "{{ protocol }}")], unit=UNITS.SHORT, w=8),
                        TimeSeriesPanel("Open connections by method", [QueryExpr(cleanup_service_label('sum(traefik_service_open_connections{namespace=~"$namespace", instance=~"$instance", service=~"$service"}[$__interval]) by (method)'), "{{ method }}")], unit=UNITS.SHORT, w=8),
                        # by service
                        TimeSeriesPanel("Request Count by service", [QueryExpr(cleanup_service_label('sum(increase(traefik_service_requests_total{namespace=~"$namespace", instance=~"$instance", service=~"$service"}[$__interval])) by (service)'), "{{ service }}")], unit=UNITS.SHORT, style="bars"),
                        TimeSeriesPanel("Request Bytes by service", [QueryExpr(cleanup_service_label('sum(increase(traefik_service_requests_bytes_total{namespace=~"$namespace", instance=~"$instance", service=~"$service"}[$__interval])) by (service)'), "{{ service }}")], unit=UNITS.BYTES, style="bars"),
                        TimeSeriesPanel("Response Bytes by service", [QueryExpr(cleanup_service_label('sum(increase(traefik_service_responses_bytes_total{namespace=~"$namespace", instance=~"$instance", service=~"$service"}[$__interval])) by (service)'), "{{ service }}")], unit=UNITS.BYTES, style="bars"),
                        TimeSeriesPanel("Avg request duration by service", [QueryExpr(cleanup_service_label(avg_duration_query("traefik_service_request_duration_seconds", '{namespace=~"$namespace", instance=~"$instance", service=~"$service"}', "service")), "{{ service }}")], unit=UNITS.SECONDS),
                        # by protocol
                        TimeSeriesPanel("Request Count by protocol", [QueryExpr(cleanup_service_label('sum(increase(traefik_service_requests_total{namespace=~"$namespace", instance=~"$instance", service=~"$service"}[$__interval])) by (protocol)'), "{{ protocol }}")], unit=UNITS.SHORT, style="bars"),
                        TimeSeriesPanel("Request Bytes by protocol", [QueryExpr(cleanup_service_label('sum(increase(traefik_service_requests_bytes_total{namespace=~"$namespace", instance=~"$instance", service=~"$service"}[$__interval])) by (protocol)'), "{{ protocol }}")], unit=UNITS.BYTES, style="bars"),
                        TimeSeriesPanel("Response Bytes by protocol", [QueryExpr(cleanup_service_label('sum(increase(traefik_service_responses_bytes_total{namespace=~"$namespace", instance=~"$instance", service=~"$service"}[$__interval])) by (protocol)'), "{{ protocol }}")], unit=UNITS.BYTES, style="bars"),
                        TimeSeriesPanel("Avg request duration by protocol", [QueryExpr(cleanup_service_label(avg_duration_query("traefik_service_request_duration_seconds", '{namespace=~"$namespace", instance=~"$instance", service=~"$service"}', "protocol")), "{{ protocol }}")], unit=UNITS.SECONDS),
                        # by method
                        TimeSeriesPanel("Request Count by method", [QueryExpr(cleanup_service_label('sum(increase(traefik_service_requests_total{namespace=~"$namespace", instance=~"$instance", service=~"$service"}[$__interval])) by (method)'), "{{ method }}")], unit=UNITS.SHORT, style="bars"),
                        TimeSeriesPanel("Request Bytes by method", [QueryExpr(cleanup_service_label('sum(increase(traefik_service_requests_bytes_total{namespace=~"$namespace", instance=~"$instance", service=~"$service"}[$__interval])) by (method)'), "{{ method }}")], unit=UNITS.BYTES, style="bars"),
                        TimeSeriesPanel("Response Bytes by method", [QueryExpr(cleanup_service_label('sum(increase(traefik_service_responses_bytes_total{namespace=~"$namespace", instance=~"$instance", service=~"$service"}[$__interval])) by (method)'), "{{ method }}")], unit=UNITS.BYTES, style="bars"),
                        TimeSeriesPanel("Avg request duration by method", [QueryExpr(cleanup_service_label(avg_duration_query("traefik_service_request_duration_seconds", '{namespace=~"$namespace", instance=~"$instance", service=~"$service"}', "method")), "{{ method }}")], unit=UNITS.SECONDS),
                        # by code
                        TimeSeriesPanel("Request Count by code", [QueryExpr(cleanup_service_label('sum(increase(traefik_service_requests_total{namespace=~"$namespace", instance=~"$instance", service=~"$service"}[$__interval])) by (code)'), "{{ code }}")], unit=UNITS.SHORT, style="bars"),
                        TimeSeriesPanel("Request Bytes by code", [QueryExpr(cleanup_service_label('sum(increase(traefik_service_requests_bytes_total{namespace=~"$namespace", instance=~"$instance", service=~"$service"}[$__interval])) by (code)'), "{{ code }}")], unit=UNITS.BYTES, style="bars"),
                        TimeSeriesPanel("Response Bytes by code", [QueryExpr(cleanup_service_label('sum(increase(traefik_service_responses_bytes_total{namespace=~"$namespace", instance=~"$instance", service=~"$service"}[$__interval])) by (code)'), "{{ code }}")], unit=UNITS.BYTES, style="bars"),
                        TimeSeriesPanel("Avg request duration by code", [QueryExpr(cleanup_service_label(avg_duration_query("traefik_service_request_duration_seconds", '{namespace=~"$namespace", instance=~"$instance", service=~"$service"}', "code")), "{{ code }}")], unit=UNITS.SECONDS),
                    ],
                ],
            ),
            *RowPanel(
                title="TLS Metrics",
                collapsed=True,
                defaultWidth=12,
                panels=[
                    [
                        TimeSeriesPanel("TLS requests", [QueryExpr('sum(increase(traefik_service_requests_tls_total{namespace=~"$namespace", instance=~"$instance"}[$__interval])) by (service, tls_cipher, version)', "{{ service }} {{ tls_cipher }} {{ version }}")], unit=UNITS.SHORT, style="bars"),
                        TimeSeriesPanel("TLS certs expiration timestamp", [QueryExpr('max(traefik_tls_certs_not_after{namespace=~"$namespace", instance=~"$instance"}) by (cn, sans, serial) * 1000', "{{ cn }} {{ sans }} {{ serial }}")], unit=UNITS.DATE_TIME_FROM_NOW),
                    ],
                ],
            ),
            *GolangResourceUsagePanels(selector='{namespace=~"$namespace", instance=~"$instance"}'),
        ]
    ),
).auto_panel_ids()

generate_dashboard(my_dashboard, "traefik.json")

File: dashboards/traefik.ts
import { CounterMetric, GaugeMetric, goRuntimeMetricsPanels, newDashboard, NewLokiDatasourceVariable, NewLokiLogsPanel, NewPanelGroup, NewPanelRow, NewPrometheusDatasourceVariable, NewQueryVariable, NewStatPanel, NewTimeSeriesPanel, PanelRowAndGroups, SummaryMetric, units } from '../src/grafana-helpers'
import { cadvisorMetricsPanels } from '../src/common-panels/k8s-cadvisor'
import * as dashboard from '@grafana/grafana-foundation-sdk/dashboard'

const datasource: dashboard.DataSourceRef = {
  uid: '${DS_PROMETHEUS}',
}

const lokiDatasource: dashboard.DataSourceRef = {
  uid: '${DS_LOKI}',
}

// https://github.com/traefik/traefik/blob/master/pkg/metrics/prometheus.go

// server meta information.
// Config reloads
const configReloads = new CounterMetric('traefik_config_reloads_total')
// Last config reload success
const lastConfigReloadSuccess = new GaugeMetric('traefik_config_last_reload_success')
// Certificate expiration timestamp
const tlsCertsNotAfterTimestamp = new GaugeMetric('traefik_tls_certs_not_after_timestamp') // cn, serial, sans
// How many open connections exist, by entryPoint and protocol
const openConnections = new GaugeMetric('traefik_open_connections') // entrypoint, protocol

// entry point.
// How many HTTP requests processed on an entrypoint, partitioned by status code, protocol, and method.
const entryPointReqs = new CounterMetric('traefik_entrypoint_requests_total') // entrypoint, code, method, protocol
// How many HTTP requests with TLS processed on an entrypoint, partitioned by TLS Version and TLS cipher Used.
const entryPointReqsTLS = new CounterMetric('traefik_entrypoint_requests_tls_total') // tls_version, tls_cipher, entrypoint
// How long it took to process the request on an entrypoint, partitioned by status code, protocol, and method.
const entryPointReqDurations = new SummaryMetric('traefik_entrypoint_request_duration_seconds') // entrypoint, code, method, protocol
// The total size of requests in bytes handled by an entrypoint, partitioned by status code, protocol, and method.
const entryPointReqsBytesTotal = new CounterMetric('traefik_entrypoint_requests_bytes_total') // entrypoint, code, method, protocol
// The total size of responses in bytes handled by an entrypoint, partitioned by status code, protocol, and method.
const entryPointRespsBytesTotal = new CounterMetric('traefik_entrypoint_responses_bytes_total') // entrypoint, code, method, protocol

// router level.
// How many HTTP requests are processed on a router, partitioned by service, status code, protocol, and method.
const routerReqs = new CounterMetric('traefik_router_requests_total') // router, service, code, method, protocol
// How many HTTP requests with TLS are processed on a router, partitioned by service, TLS Version, and TLS cipher Used.
const routerReqsTLS = new CounterMetric('traefik_router_requests_tls_total') // tls_version, tls_cipher, router, service
// How long it took to process the request on a router, partitioned by service, status code, protocol, and method.
const routerReqDurations = new SummaryMetric('traefik_router_request_duration_seconds') // router, service, code, method, protocol
// The total size of requests in bytes handled by a router, partitioned by service, status code, protocol, and method.
const routerReqsBytesTotal = new CounterMetric('traefik_router_requests_bytes_total') // router, service, code, method, protocol
// The total size of responses in bytes handled by a router, partitioned by service, status code, protocol, and method.
const routerRespsBytesTotal = new CounterMetric('traefik_router_responses_bytes_total') // router, service, code, method, protocol

// service level.
// How many HTTP requests processed on a service, partitioned by status code, protocol, and method.
const serviceReqs = new CounterMetric('traefik_service_requests_total') // service, code, method, protocol
// How many HTTP requests with TLS processed on a service, partitioned by TLS version and TLS cipher.
const serviceReqsTLS = new CounterMetric('traefik_service_requests_tls_total') // tls_version, tls_cipher, service
// How long it took to process the request on a service, partitioned by status code, protocol, and method.
const serviceReqDurations = new SummaryMetric('traefik_service_request_duration_seconds') // service, code, method, protocol
// How many request retries happened on a service.
const serviceRetries = new CounterMetric('traefik_service_retries_total') // service
// service server is up, described by gauge value of 0 or 1.
const serviceServerUp = new GaugeMetric('traefik_service_server_up') // service, url
// The total size of requests in bytes received by a service, partitioned by status code, protocol, and method.
const serviceReqsBytesTotal = new CounterMetric('traefik_service_requests_bytes_total') // service, code, method, protocol
// The total size of responses in bytes returned by a service, partitioned by status code, protocol, and method.
const serviceRespsBytesTotal = new CounterMetric('traefik_service_responses_bytes_total') // service, code, method, protocol

const selectors = `namespace=~"$namespace", instance=~"$instance"`

const serviceMetricsPanels = (opts: { title?: string; groupBy: string[]; extraSelectors?: string[]; append?: string }) => {
  const { title = 'Service Metrics', extraSelectors = [], append, groupBy } = opts
  return NewPanelGroup({ title: `${title} (by ${opts.groupBy.join(', ')})` }, [
    // available labels: service, code, method, protocol
    NewPanelRow({ datasource, height: 8 }, [
      // by service
      NewTimeSeriesPanel({ title: `Request Count by ${opts.groupBy.join(', ')}` }, serviceReqs.calc('sum', 'increase', { selectors: [selectors, 'service=~"$service"', ...extraSelectors], groupBy, append }).target()),
      NewTimeSeriesPanel({ title: `Request Bytes by ${opts.groupBy.join(', ')}` }, serviceReqsBytesTotal.calc('sum', 'increase', { selectors: [selectors, 'service=~"$service"', ...extraSelectors], groupBy, append }).target()),
      NewTimeSeriesPanel({ title: `Response Bytes by ${opts.groupBy.join(', ')}` }, serviceRespsBytesTotal.calc('sum', 'increase', { selectors: [selectors, 'service=~"$service"', ...extraSelectors], groupBy, append }).target()),
      NewTimeSeriesPanel({ title: `Avg request duration by ${opts.groupBy.join(', ')}` }, serviceReqDurations.avg({ selectors: [selectors, 'service=~"$service"', ...extraSelectors], groupBy, append }).target()),
    ]),
  ])
}

const panels: PanelRowAndGroups = [
  NewPanelGroup({ title: 'Overview' }, [
    NewPanelRow({ datasource, height: 3 }, [
      NewStatPanel({ title: 'Request Count' }, serviceReqs.calc('sum', 'increase', { selectors: [selectors, 'service=~"$service"'], interval: '$__range' }).target()),
      NewStatPanel({ title: 'Request Bytes' }, serviceReqsBytesTotal.calc('sum', 'increase', { selectors: [selectors, 'service=~"$service"'], interval: '$__range' }).target()),
      NewStatPanel({ title: 'Response Bytes' }, serviceRespsBytesTotal.calc('sum', 'increase', { selectors: [selectors, 'service=~"$service"'], interval: '$__range' }).target()),
      NewStatPanel({ title: 'Config reloads' }, configReloads.calc('sum', 'increase', { selectors, interval: '$__range' }).target()),
      NewStatPanel({ title: 'Last successful config reload', unit: units.DateTimeFromNow }, lastConfigReloadSuccess.calc('max', { selectors, append: '* 1000' }).target()),
    ]),
    NewPanelRow({ datasource, height: 8 }, [
      //
      NewTimeSeriesPanel({ title: 'Open connections' }, openConnections.calc('sum', { selectors: [selectors, 'entrypoint=~"$entrypoint"'], groupBy: ['entrypoint', 'protocol'] }).target()),
    ]),
  ]),
  NewPanelGroup({ title: 'Entrypoint Metrics' }, [
    // available labels: entrypoint, code, method, protocol
    NewPanelRow({ datasource, height: 8 }, [
      // by entrypoint
      NewTimeSeriesPanel({ title: 'Request Count' }, entryPointReqs.calc('sum', 'increase', { selectors: [selectors, 'entrypoint=~"$entrypoint"'], groupBy: ['entrypoint'] }).target()),
      NewTimeSeriesPanel({ title: 'Request Bytes' }, entryPointReqsBytesTotal.calc('sum', 'increase', { selectors: [selectors, 'entrypoint=~"$entrypoint"'], groupBy: ['entrypoint'] }).target()),
      NewTimeSeriesPanel({ title: 'Response Bytes' }, entryPointRespsBytesTotal.calc('sum', 'increase', { selectors: [selectors, 'entrypoint=~"$entrypoint"'], groupBy: ['entrypoint'] }).target()),
      NewTimeSeriesPanel({ title: 'Avg request duration' }, entryPointReqDurations.avg({ selectors: [selectors, 'entrypoint=~"$entrypoint"'], groupBy: ['entrypoint'] }).target()),
    ]),
  ]),
  serviceMetricsPanels({ groupBy: ['service'] }),
  serviceMetricsPanels({ groupBy: ['protocol'] }),
  serviceMetricsPanels({ groupBy: ['method'] }),
  serviceMetricsPanels({ groupBy: ['code'] }),
  serviceMetricsPanels({ title: 'Service Metrics (errors)', groupBy: ['code', 'protocol', 'method', 'service'], extraSelectors: ['code=~"(4|5).."'], append: ' > 0' }),
  // NewPanelGroup({ title: 'Service Metrics (by code) (errors)' }, [
  //   // NewPanelRow({ datasource, height: 8 }, [
  //   //   //
  //   //   NewTimeSeriesPanel({ title: 'Service Retries' }, serviceRetries.calc('sum', 'increase', { selectors: [selectors, 'service=~"$service"'], groupBy: ['service'] }).target()),
  //   //   NewTimeSeriesPanel({ title: 'Service Server Up' }, serviceServerUp.calc('max', { selectors: [selectors, 'service=~"$service"'], groupBy: ['service', 'url'] }).target()),
  //   // ]),
  // ]),
  NewPanelGroup({ title: 'TLS Metrics' }, [
    NewPanelRow({ datasource, height: 8 }, [
      //
      NewTimeSeriesPanel({ title: 'TLS requests by entrypoint' }, entryPointReqsTLS.calc('sum', 'increase', { selectors: [selectors, 'entrypoint=~"$entrypoint"'], groupBy: ['entrypoint', 'tls_cipher', 'version'] }).target()),
      NewTimeSeriesPanel({ title: 'TLS requests by service' }, serviceReqsTLS.calc('sum', 'increase', { selectors: [selectors, 'service=~"$service"'], groupBy: ['service', 'tls_cipher', 'version'] }).target()),
      NewTimeSeriesPanel({ title: 'TLS certs expiration timestamp', unit: units.DateTimeFromNow }, tlsCertsNotAfterTimestamp.calc('max', { selectors: [selectors], groupBy: ['cn', 'sans', 'serial'], append: ' * 1000' }).target()),
    ]),
  ]),
  NewPanelGroup({ title: 'Logs' }, [
    // {container_name="traefik"} | json | DownstreamStatus >= 400
    // ClientAddr RequestAddr RequestPath DownstreamStatus RouterName
    // 'ServiceAddr', 'ServiceName', 'ServiceURL', 'level', 'Duration', 'OriginDuration'
    NewPanelRow({ datasource: lokiDatasource, height: 16 }, [NewLokiLogsPanel({ title: 'Downstream errors (4xx)' }, { expr: `{container_name="traefik"} | json | DownstreamStatus >= 400 and DownstreamStatus < 500 | line_format "{{.ClientAddr}} - {{.RequestProtocol}} - {{.RequestScheme}}://{{.RequestAddr}}{{.RequestPath}} - {{.RouterName}} - {{.DownstreamStatus}} - {{.OriginStatus}}"` })]),
    NewPanelRow({ datasource: lokiDatasource, height: 16 }, [NewLokiLogsPanel({ title: 'Downstream errors (5xx)' }, { expr: `{container_name="traefik"} | json | DownstreamStatus >= 500 | line_format "{{.ClientAddr}} - {{.RequestProtocol}} - {{.RequestScheme}}://{{.RequestAddr}}{{.RequestPath}} - {{.RouterName}} - {{.DownstreamStatus}} - {{.OriginStatus}}"` })]),
    // NewPanelRow({ datasource: lokiDatasource, height: 16 }, [NewLokiLogsPanel({ title: 'Downstream errors (0)' }, { expr: `{container_name="traefik"} | json | DownstreamStatus == 0 | line_format "{{.ClientAddr}} - {{.RequestProtocol}} - {{.RequestScheme}}://{{.RequestAddr}}{{.RequestPath}} - {{.RouterName}} - {{.DownstreamStatus}} - {{.OriginStatus}}"` })]),
  ]),
  goRuntimeMetricsPanels({ datasource, selectors, collapsed: true }),
  cadvisorMetricsPanels({ datasource, selectors: [`namespace=~"$namespace"`, `pod=~"$pod"`], collapsed: true }), //`container="traefik"`
]

export type LokiLogsPanelOpts = {
  title: string
  query: string
  limit: number
}

export const traefikDashboard = newDashboard({
  variables: [
    NewPrometheusDatasourceVariable({ name: 'DS_PROMETHEUS', label: 'Prometheus' }),
    NewLokiDatasourceVariable({ name: 'DS_LOKI', label: 'Loki' }),
    NewQueryVariable({ datasource, name: 'namespace', label: 'Namespace', query: 'label_values(traefik_config_reloads_total, namespace)', includeAll: true, multi: true }),
    NewQueryVariable({ datasource, name: 'instance', label: 'Instance', query: 'label_values(traefik_config_reloads_total{namespace=~"$namespace"}, instance)', includeAll: true, multi: true }),
    NewQueryVariable({ datasource, name: 'pod', label: 'Pod', query: 'label_values(traefik_config_reloads_total{namespace=~"$namespace", instance=~"$instance"}, pod)', includeAll: true, multi: true }),
    NewQueryVariable({ datasource, name: 'entrypoint', label: 'Entrypoint', query: 'label_values(traefik_entrypoint_requests_total{instance=~"$instance"}, entrypoint)', includeAll: true, multi: true }),
    NewQueryVariable({ datasource, name: 'service', label: 'Service', query: 'label_values(traefik_service_requests_total{instance=~"$instance"}, service)', includeAll: true, multi: true }),
  ],
  panels,
})
